File: backend/app/clients/datagov.py
import logging
from typing import Optional

from .base import BaseAPIClient
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class DataGovClient(BaseAPIClient):
    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.datagov_base_url)
        self.api_key = settings.gov_api_key

    def _get_headers(self) -> dict:
        headers = {"Accept": "application/json"}
        if self.api_key:
            headers["X-Api-Key"] = self.api_key
        return headers

    def _normalize_dataset(self, dataset: dict) -> SourceItem:
        resources = dataset.get("resources", [])
        pdf_url = None
        for resource in resources:
            if resource.get("format", "").upper() == "PDF":
                pdf_url = resource.get("url")
                break

        organization = dataset.get("organization", {})
        agency = organization.get("title") if organization else None

        dataset_id = dataset.get("id") or dataset.get("name", "")
        url = f"https://catalog.data.gov/dataset/{dataset_id}"

        return SourceItem(
            source_type="datagov",
            id=dataset_id,
            title=dataset.get("title", "Untitled Dataset"),
            agency=agency,
            date=dataset.get("metadata_modified") or dataset.get("metadata_created"),
            url=url,
            excerpt=dataset.get("notes", "")[:500] if dataset.get("notes") else None,
            pdf_url=pdf_url,
            content_type="dataset",
            raw=dataset,
        )

    async def search_datasets(
        self,
        query: str,
        organization: Optional[str] = None,
        groups: Optional[list[str]] = None,
        res_format: Optional[str] = None,
        rows: int = 10,
        start: int = 0,
    ) -> tuple[list[dict], list[SourceItem]]:
        fq_parts = []
        if organization:
            fq_parts.append(f'organization:"{organization}"')
        if groups:
            for group in groups:
                fq_parts.append(f'groups:"{group}"')
        if res_format:
            fq_parts.append(f'res_format:"{res_format}"')

        params = {
            "q": query,
            "rows": rows,
            "start": start,
        }

        if fq_parts:
            params["fq"] = " AND ".join(fq_parts)

        url = f"{self.base_url}/package_search"
        logger.info(f"Searching data.gov: {query}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        result = data.get("result", {})
        datasets = result.get("results", [])
        sources = [self._normalize_dataset(ds) for ds in datasets]

        return datasets, sources

File: backend-ts/src/clients/datagov.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";

export class DataGovClient extends BaseAPIClient {
    private apiKey: string;

    constructor() {
        const settings = getSettings();
        super(settings.dataGovBaseUrl);
        this.apiKey = settings.govApiKey;
    }

    private getHeaders(): Record<string, string> {
        const headers: Record<string, string> = { Accept: "application/json" };
        if (this.apiKey) {
            headers["X-Api-Key"] = this.apiKey;
        }
        return headers;
    }

    private normalizeDataset(dataset: Record<string, unknown>): SourceItem {
        const resources = (dataset.resources || []) as { format?: string; url?: string }[];
        let pdfUrl: string | null = null;

        for (const resource of resources) {
            if ((resource.format || "").toUpperCase() === "PDF" && resource.url) {
                pdfUrl = resource.url;
                break;
            }
        }

        const organization = dataset.organization as Record<string, unknown> | undefined;
        const agency = organization?.title as string || null;

        const datasetId = (dataset.id as string) || (dataset.name as string) || "";
        const url = `https://catalog.data.gov/dataset/${datasetId}`;

        const notes = dataset.notes as string | undefined;

        return {
            source_type: "datagov",
            id: datasetId,
            title: (dataset.title as string) || "Untitled Dataset",
            agency,
            date:
                (dataset.metadata_modified as string) ||
                (dataset.metadata_created as string) ||
                null,
            url,
            excerpt: notes ? notes.slice(0, 500) : null,
            pdf_url: pdfUrl,
            content_type: "dataset",
            raw: dataset,
        };
    }

    async searchDatasets(options: {
        query: string;
        organization?: string;
        groups?: string[];
        resFormat?: string;
        rows?: number;
        start?: number;
    }): Promise<{ datasets: Record<string, unknown>[]; sources: SourceItem[] }> {
        const { query, organization, groups, resFormat, rows = 10, start = 0 } = options;

        const fqParts: string[] = [];
        if (organization) {
            fqParts.push(`organization:"${organization}"`);
        }
        if (groups && groups.length > 0) {
            for (const group of groups) {
                fqParts.push(`groups:"${group}"`);
            }
        }
        if (resFormat) {
            fqParts.push(`res_format:"${resFormat}"`);
        }

        const params: Record<string, string | number> = {
            q: query,
            rows,
            start,
        };

        if (fqParts.length > 0) {
            params.fq = fqParts.join(" AND ");
        }

        const url = `${this.baseUrl}/package_search`;
        console.log(`Searching data.gov: ${query}`);

        const data = await this.requestWithRetry<{
            result: { results: Record<string, unknown>[] };
        }>({
            url,
            headers: this.getHeaders(),
            params,
        });

        const datasets = data.result?.results || [];
        const sources = datasets.map((ds) => this.normalizeDataset(ds));

        return { datasets, sources };
    }
}
